File: probe/src/main/kotlin/schemagraph/probe/DocumentVersions.kt
package schemagraph.probe

import com.fasterxml.jackson.databind.node.ObjectNode

/** 필드 변경은 전송 경계에서만 처리해 추출기 내부 모델과 v1 호환성을 유지한다. */
internal fun documentWire(doc: CatalogDocument, version: Int): ObjectNode {
    require(version in 1..2) { "Unsupported catalog version; choose 1 or 2" }
    require(version == 1 || doc.reader.isNotEmpty()) { "Catalog v2 requires a nonempty producer name" }
    val node = mapper.valueToTree<ObjectNode>(doc)
    node.put("version", version)
    if (version == 2) {
        node.remove("reader")
        node.set<ObjectNode>("producer", mapper.createObjectNode().put("name", doc.reader))
        node.set<com.fasterxml.jackson.databind.JsonNode>("required_features", mapper.valueToTree(documentFeatures(doc)))
    }
    return node
}

/** 필수 기능은 실제 레코드의 의미를 기준으로 선언해 미지원 소비자의 무시를 막는다. */
private fun documentFeatures(doc: CatalogDocument): List<String> {
    val features = sortedSetOf<String>()
    if (doc.dependencies.isNotEmpty()) features += "catalog-dependencies-v1"
    for (schema in doc.schemas) {
        if (schema.objects.any { it.usage != null || it.indexes.any { index -> index.usage != null } }
            || schema.routines.any { it.usage != null }) features += "usage-v1"
        if (schema.routines.any { it.memberOf != null }) features += "package-members-v1"
    }
    return features.toList()
}

/** 스트리밍은 레코드를 보기 전에 헤더를 내보내므로 방언이 수확할 수 있는 기능을 선언한다. */
internal fun streamingHeader(version: Int, dialect: String, catalogDependencies: Boolean = false): ObjectNode {
    require(version in 1..2) { "Unsupported catalog version; choose 1 or 2" }
    val node = lineMapper.createObjectNode()
        .put("type", "document").put("version", version).put("dialect", dialect)
    node.set<com.fasterxml.jackson.databind.JsonNode>("limitations", lineMapper.valueToTree(emptyList<String>()))
    if (version == 1) {
        node.put("reader", "probe-jdbc")
    } else {
        node.set<ObjectNode>("producer", lineMapper.createObjectNode().put("name", "probe-jdbc"))
        val features = (when (dialect) {
            "postgres", "mysql", "mariadb" -> listOf("usage-v1")
            "oracle" -> listOf("package-members-v1")
            else -> emptyList()
        } + if (catalogDependencies) listOf("catalog-dependencies-v1") else emptyList()).sorted()
        node.set<com.fasterxml.jackson.databind.JsonNode>("required_features", lineMapper.valueToTree(features))
    }
    return node
}

File: probe-go/document_versions.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"
)

// validateDocumentVersion은 송신자가 엔진이 정의한 문서 envelope만 만들도록
// 버전을 먼저 제한한다. 알 수 없는 버전은 필드를 추측해 내보내지 않는다.
func validateDocumentVersion(version int) error {
	if version != 1 && version != 2 {
		return fmt.Errorf("unsupported document version %d; choose 1 or 2", version)
	}
	return nil
}

type documentProducer struct {
	Name string `json:"name"`
}

type documentV2 struct {
	Version          int                 `json:"version"`
	Dialect          string              `json:"dialect"`
	Producer         documentProducer    `json:"producer"`
	RequiredFeatures []string            `json:"required_features"`
	Schemas          []SchemaDoc         `json:"schemas"`
	Limitations      []string            `json:"limitations"`
	Context          *CollectionContext  `json:"context,omitempty"`
	Dependencies     []CatalogDependency `json:"dependencies,omitempty"`
}

type ndjsonDocumentV1 struct {
	Type        string             `json:"type"`
	Version     int                `json:"version"`
	Dialect     string             `json:"dialect"`
	Reader      string             `json:"reader"`
	Limitations []string           `json:"limitations"`
	Context     *CollectionContext `json:"context,omitempty"`
}

type ndjsonDocumentV2 struct {
	Type             string             `json:"type"`
	Version          int                `json:"version"`
	Dialect          string             `json:"dialect"`
	Producer         documentProducer   `json:"producer"`
	RequiredFeatures []string           `json:"required_features"`
	Limitations      []string           `json:"limitations"`
	Context          *CollectionContext `json:"context,omitempty"`
}

// encodeDocument은 JSON과 NDJSON 모두에 같은 v1/v2 envelope 규칙을 적용한다.
// 레코드 data는 typed 구조체로 직렬화해 큰 int64 usage 카운터를 보존한다.
func encodeDocument(doc CatalogDocument, version int, format string) ([]byte, error) {
	if err := validateDocumentVersion(version); err != nil {
		return nil, err
	}
	if version == 2 && doc.Reader == "" {
		return nil, fmt.Errorf("catalog v2 requires a nonempty producer name")
	}
	switch format {
	case "json":
		return encodeDocumentJSON(doc, version)
	case "ndjson":
		return encodeDocumentNDJSON(doc, version)
	default:
		return nil, fmt.Errorf("unsupported document format %q; choose json or ndjson", format)
	}
}

func encodeDocumentJSON(doc CatalogDocument, version int) ([]byte, error) {
	var value any
	if version == 1 {
		v1 := doc
		v1.Version = version
		value = v1
	} else {
		value = documentV2{
			Version: version, Dialect: doc.Dialect,
			Producer:         documentProducer{Name: doc.Reader},
			RequiredFeatures: documentFeatures(doc),
			Schemas:          doc.Schemas, Limitations: doc.Limitations,
			Context: doc.Context, Dependencies: doc.Dependencies,
		}
	}
	encoded, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return nil, err
	}
	return append(encoded, '\n'), nil
}

func encodeDocumentNDJSON(doc CatalogDocument, version int) ([]byte, error) {
	var b strings.Builder
	line := func(value any) error {
		encoded, err := json.Marshal(value)
		if err != nil {
			return err
		}
		b.Write(encoded)
		b.WriteByte('\n')
		return nil
	}

	if version == 1 {
		if err := line(ndjsonDocumentV1{
			Type: "document", Version: version, Dialect: doc.Dialect,
			Reader: doc.Reader, Limitations: []string{},
			Context: doc.Context,
		}); err != nil {
			return nil, err
		}
	} else if err := line(ndjsonDocumentV2{
		Type: "document", Version: version, Dialect: doc.Dialect,
		Producer:         documentProducer{Name: doc.Reader},
		RequiredFeatures: documentFeatures(doc), Limitations: []string{},
		Context: doc.Context,
	}); err != nil {
		return nil, err
	}

	for _, schema := range doc.Schemas {
		if err := line(map[string]any{"type": "schema", "name": schema.Name}); err != nil {
			return nil, err
		}
		for _, object := range schema.Objects {
			if err := line(map[string]any{
				"type": "object", "schema": schema.Name, "data": object,
			}); err != nil {
				return nil, err
			}
		}
		for _, routine := range schema.Routines {
			if err := line(map[string]any{
				"type": "routine", "schema": schema.Name, "data": routine,
			}); err != nil {
				return nil, err
			}
		}
	}
	for _, dependency := range doc.Dependencies {
		if err := line(map[string]any{"type": "dependency", "data": dependency}); err != nil {
			return nil, err
		}
	}
	if err := line(map[string]any{"type": "limitations", "data": doc.Limitations}); err != nil {
		return nil, err
	}
	return []byte(b.String()), nil
}

func documentFeatures(doc CatalogDocument) []string {
	usage, members := false, false
	for _, schema := range doc.Schemas {
		for _, object := range schema.Objects {
			usage = usage || object.Usage != nil
			for _, index := range object.Indexes {
				usage = usage || index.Usage != nil
			}
		}
		for _, routine := range schema.Routines {
			usage = usage || routine.Usage != nil
			members = members || routine.MemberOf != nil
		}
	}
	features := make([]string, 0, 2)
	if len(doc.Dependencies) > 0 {
		features = append(features, "catalog-dependencies-v1")
	}
	if members {
		features = append(features, "package-members-v1")
	}
	if usage {
		features = append(features, "usage-v1")
	}
	return features
}
